### DISCUSSION TITLE
question on validation rules (pattern)

### DISCUSSION THREAD

**Discussion by Elfayer** [react-hook-form/react-hook-form#7228] [Q&A] [answered]
### Version Number

7

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-v7-controller-forked-3dqvn?file=/src/Mui.js

### Steps to reproduce

1. Go to https://codesandbox.io/s/react-hook-form-v7-controller-forked-3dqvn?file=/src/Mui.js
2. Type `123.123` in the MUI TextField
3. Submit
4. See there are no errors on the console from the MUI Textfield

As you can see the error works for a simple `<input/>`. It's as if the `pattern` was not checked.

### Expected behaviour

The `pattern` should put the MUI Textfield to error.

If I replace the `onChange` from:
```js
onChange={(e) => {
  const { value, valueAsNumber } = e.target;

  if (!isNaN(valueAsNumber)) {
    field.onChange(valueAsNumber);
  } else {
    field.onChange(value);
  }
}}
```
to:
```js
onChange={(e) => {
  const { value } = e.target;

  field.onChange(value);
}}
```
The validation works. So it has to come from `field.onChange(valueAsNumber);`.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment, marked as answer]
```tsx
<input
  type="number"
  step="0.001"
  onChange={(e) => {
    const { value, valueAsNumber } = e.target;

    if (!isNaN(valueAsNumber)) {
      setValue("text", valueAsNumber);
    } else {
      setValue("text", value);
    }
  }} // no effect
  {...register("text", {
    required: true,
    pattern: /^\d+(?:\.\d{1,2})?$/,
    min: 0.01
  })}
/>
```

look at the code, the onChange for native input is getting overwritten by what's returned by `register`, I would first double-check your regex.

## Pattern

the regex pattern right now only applied to `string` as it following the standard `pattern` attribute from native input, if you need to validate against number, you can use `validate` function by convert it back to `string`

**Reply by Elfayer** [reply]
Wow that was fast. Thanks for the reply!
Sorry my bad for that part.^^
So they both behave the same, the `onChange` is probably wrong, yet I don't see what I should do to fix it.
The regex looks good:
```js
/^\d+(?:\.\d{1,2})?$/.test(12.12)
true

/^\d+(?:\.\d{1,2})?$/.test(12.123)
false

/^\d+(?:\.\d{1,2})?$/.test("12.12")
true

/^\d+(?:\.\d{1,2})?$/.test("12.123")
false
```

**Reply by bluebill1049** [reply]
no worries, but we are following the standard here which is only apply to `string`, also if you are using typescript, it probably not going to compile as well.

**Reply by Elfayer** [reply]
Noted. Do you have a suggestion to store numbers in the form data?

Is this also expected? The following won't raise an error (I guess it goes the same as I manually wrote 🤔 )
```js
<input
  type="number"
  step="0.001"
  {...register("text", {
    valueAsNumber: true, // added
    required: true,
    pattern: /^\d+(?:\.\d{1,2})?$/, // not called
    min: 0.01
  })}
/>
```

**Reply by bluebill1049** [reply]
use `validate` function instead, cause it's converted to `number`.

**Reply by Elfayer** [reply]
Thanks for being very efficient! ❤️

**Reply by bluebill1049** [reply]
you are welcome.
